[sailfish-connman-plugin-iptables] First version of sailfish connman iptables management plugin providing privileged users a D-Bus API for managing iptables content. Contributes to JB#39338

LaakkonenJussi opened the pull request:
Features implemented in this version:

This plugin allows to:
 - Add a rule to iptables filter table
 - Remove a rule from iptables filter table
 - Change policy of a filter table chain INPUT and OUTPUT
 - Add a custom chain to filter table
 - Clear iptables rules
 - Clear iptables custom chains
 - Get iptables filter table content
 - Get version of plugin interface
 - Register (and unregister) to listen for API change signals
 
Rules can be added to any chains found iptables filter table. Each rule
can be added with any iptables target: ACCEPT, DENY, REJECT, QUEUE, LOG and
custom chains can be used as targets too. Custom chains must be used with the
name they are added, e.g., adding chain CUSTOM1 is added as sfos_CUSTOM1 and it
has to be used without the prefix "sfos_" as CUSTOM1.

Following parameters are supported:
 - Ip address or network 
 - Ip address or network with port
 - Ip address or network with port range
 - Ip address or network with service name
 - Port with any address 0.0.0.0
 - Port range with any address 0.0.0.0
 - Service name with any address 0.0.0.0
 
Both source and destination parameters are supported for all. Either or both
parameters (source or destination) must be set. They are added to iptables
accordingly.
 
For more information refer to README.md or xml documentation of the D-Bus interface of this
plugin (sailfish_iptables_dbus_interface_description.xml).

LaakkonenJussi committed:
Initial commit

LaakkonenJussi committed:
[sailfish-connman-plugin-iptables] Added package spec and makefile.

LaakkonenJussi committed:
[sailfish-connman-plugin-iptables] First version of sailfish iptables plugin for connman.

Features implemented in this version:

Add a rule to iptables filter table
Remove a rule from iptables filter table
Change policy of a filter table chain INPUT and OUTPUT
Save firewall (all iptables content) to disk
Clear firewall (all iptables tables)
Get version of plugin interface

Support for IP address, network, port, port range and service name is implemented.

LaakkonenJussi committed:
Merge remote-tracking branch 'origin' into jb39338

LaakkonenJussi committed:
[sailfish-connman-plugin-iptables] Updated documentation.

spiiroin commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+	DBusMessage* reply = sailfish_iptables_dbus_method_return(message,
+			DBUS_TYPE_UINT16,	(guint16)result,
+			DBUS_TYPE_STRING, 	api_result_message(result),
+			DBUS_TYPE_INVALID);
```
Unless I'm missing something, all of these are wrong...  should pass pointers, not values. (and should use dbus types, not glib ones)

LaakkonenJussi replied:
Fixed in 257c056893e8799c614ee7d78d2741dfb28bdc66

spiiroin commented on src/sailfish-iptables-dbus.c:
```diff
@@ -0,0 +1,696 @@
+	7 = invalid file path
+	8 = cannot process rule
+	9 = cannot perform remove operation (rule does not exist)
+*/
```
it is... "typedef enum sailfish_iptables_result" -> might refer to it?

LaakkonenJussi replied:
Added some note in 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables-dbus.c:
```diff
@@ -0,0 +1,696 @@
+void sailfish_iptables_dbus_send_signal(DBusMessage *signal)
+{
+	DBusConnection* connman_dbus = dbus_connection_ref(
+			connman_dbus_get_connection());
```
Yes, https://git.merproject.org/mer-core/connman/blob/master/connman/src/dbus.c#L688

monich replied:
It still makes no sense to grab the reference just for the duration of `g_dbus_send_message` call. It's harmless though.

LaakkonenJussi replied:
Right, the dbus_connection_ref() is already done in connman_dbus_get_connection() (https://git.merproject.org/slava/connman/blob/datacounters/connman/src/dbus.c#L688). It is, however, necessary to unref the connection. Fixed in 3b061fba4f3d01a1dba3363b67c02dcc04299b40

LaakkonenJussi committed:
[sailfish-connman-plugin-iptables] Updated to use changed headers of connman. Cleanup and various fixes. License to BSD-3.

Fixed dbus message and signal creation.

Simplified code by using dbus/g_dbus functions.

Other various cleanups to make code more readable.

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+	return __add_rule_to_iptables(params,OPERATION_OUT | OPERATION_DENY);
+}
+
+DBusMessage* __process_request(DBusMessage *message,
```
Fixed 725b7511863983275f33db5f9e8186817cb9075f.

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+					result);
+		}
+	
+		if(!result)
```
Fixed 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+				DBG("%s %s %s", "__set_policy(): changed policy", 
+					ipt_operation, params->policy);
+			}	
+			rval = OK;
```
Fixed 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+	api_result rval = OK;
+	const gchar* ipt_operation = NULL;
+	
+	if(params && (rval = check_parameters(params)) == OK)
```
Fixed 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+		
+		if(tokens)
+		{
+			for(index = 0; tokens[index] && index < 2; index++)
```
Fixed 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+			break;
+	}
+	
+	if(error)
```
Fixed 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+	
+	gchar *ip = NULL, *service = NULL, *protocol = NULL, *port_str = NULL;
+	gchar *path = NULL, *table = NULL, *policy = NULL, *operation = NULL;
+	guint16 port[2] = {0};
```
Fixed 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+		if(params->table)
+			g_free(params->table);
+		if(params->policy)
+			g_free(params->policy);
```
Simplified in 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+
+static gboolean validate_policy(const gchar* policy)
+{
+	if(policy && strlen(policy))
```
Replaced all strlen()s in 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+	{
+		operation_uppercase = g_utf8_strup(operation,-1);
+
+		if(!g_strcmp0(operation_uppercase,"ADD"))
```
Changed to use g_ascii_strcasecmp() in 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+		
+		// Allow negation
+		if(negated_ip_address(ip))
+			address = &(ip[1]);
```
Cleaned in 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+		// Dot notation mask
+		if(strchr(mask,'.'))
+			return validate_address(type, mask);
+		else
```
Cleaned 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+		
+	ip_and_mask = g_strsplit(ip,IP_MASK_DELIM,2);
+
+	if(ip_and_mask && g_strv_length(ip_and_mask) == 2)
```
Added null check to everywhere g_strsplit() is used 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+		return mask;
+
+	// Value between protocol max and 2^32, calculate cidr mask
+	else
```
Cleaned 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables-dbus.h:
```diff
@@ -0,0 +1,182 @@
+ */
+
+#ifndef __SAILFISH_IPTABLES_DBUS_H_
+#define __SAILFISH_IPTABLES_DBUS_H_
```
Removed all double underscores 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables-dbus.c:
```diff
@@ -0,0 +1,696 @@
+#define SAILFISH_IPTABLES_ALLOW_IN_IP				"AllowIncomingIp"
+#define SAILFISH_IPTABLES_ALLOW_IN_IP_PORT			"AllowIncomingIpWithPort"
+#define SAILFISH_IPTABLES_ALLOW_IN_IP_PORT_RANGE	"AllowIncomingIpWithPortRange"
+#define SAILFISH_IPTABLES_ALLOW_IN_PORT				"AllowIncomingPort"
```
Fixed in 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on rpm/sailfish-iptables-api-plugin.spec:
```diff
@@ -0,0 +1,46 @@
+%setup -q -n %{name}-%{version}
+
+%build
+make %{?jobs:-j%jobs} release
```
Change made in 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on Makefile:
```diff
@@ -0,0 +1,127 @@
+
+INSTALL_PERM  = 755
+INSTALL_OWNER = $(shell id -u)
+INSTALL_GROUP = $(shell id -g)
```
Removed in 725b7511863983275f33db5f9e8186817cb9075f

LaakkonenJussi commented on src/sailfish-iptables.c:
```diff
@@ -0,0 +1,1241 @@
+
+static gboolean validate_port(guint16 port)
+{
+	return port && port < 0xFFFF;
```
Well if the input is 0xdeadbeef to function where parameters is unsigned 16bit int isn't the input just 0xbeef ? :)

LaakkonenJussi committed:
[sailfish-connman-plugin-iptables] Removed save and load of iptables from D-Bus. Plugin saves and loads the firewall at shutdown & startup.

Upgraded required version of connman.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] XML specification to "spec" dir.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Restructuring of code for unit test needs.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Added automated unit tests. Added execution of tests to spec file.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Fixes to issues in code that were found with unit tests.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Updated XML specification. More restructuring of code.

Also minor fixes to input checks.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Added a bash script as test tool for the iptables D-Bus API.

The test can be run only on the device. Requires connman and sailfish-connman-iptables-plugin to be running.

Adds rules to iptables using the D-Bus API, checks that rule exists in iptables and removes the rule from iptables. Executes one add and remove operation for each interface.

Quits if one test fails and returns 1. Otherwise 0 if all tests are success.

monich commented on Makefile:
```diff
@@ -0,0 +1,127 @@
+ADD_LDFLAGS=`pkg-config --libs glib-2.0 dbus-1`
+
+FULL_CFLAGS = $(BASE_FLAGS) $(CFLAGS) $(ADD_CFLAGS) $(DEFINES) $(WARNINGS) -MMD -MP
+FULL_LDFLAGS = $(BASE_FLAGS) $(LDFLAGS) $(ADD_LDFLAGS) -shared -Wl,-soname -Wl,$(LIB_SONAME)
```
No it doesn't, this can (and should) be removed.

LaakkonenJussi replied:
Removed in 9a7bb06d6940f38a756df8d23ea526779cf90e46

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Packaging for unit tests + running script and end-to-end test.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Improved test script manageability.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Removed -soname from LDFLAGs.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Build fixes.

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,464 @@
+Following methods take following parameters as input:
+ NAME		TYPE	DESCRIPTION
+ ip		s		IP address or network with CIDR or dot notation (1.2.3.4/24
+ 					or 1.2.3.4/255.255.255.0)
```
IPv4 or IPv6? Number of mask bits should be a separate parameter, I think. Also, it makes every bit of sense to send IP address as a byte array to avoid ambiguities associated with text representation (like `012.0.0.1` vs `10.0.0.1` - is it the same address or not? not to mention various kinds of IPv6 representation). If both IPv4 and IPv6 are supposed to be supported, I would add address family as a separate parameter (although some could argue that one can tell IPv4 from IPv6 based on number of bytes in the address).

LaakkonenJussi replied:
Only IPv4 is currently required, some parts were thought IPv6 in mind, excluding the API. 

Here is a following problem with byte-encoded IP-addresses: The app using the API has to use some lists of byte-encoded addresses to be sent to the API, which are not human readable, thus, such data is not feasible to be added into a configuration file. Binary storage of the addresses works of course, but the content is not directly editable. Each edit would require a conversion from byte-encoded format to text format. Same thing happens with the API use: the application using the API would send the IP addresses as byte-encoded and the validity can be checked if the data is compatible with `sockaddr_in/sockaddr_in6` structures.

The sockaddr_in contains address as `uint32`, whereas in `sockaddr_in6` address is `unsigned char[16]`. It would, therefore need separate methods. Unless we want the `sockaddr_storage`structure to be sent, which may result in taking care of  the struct.

I think here a standard way is the best one, so either support the address as string or separate methods for the two IP types. 

Eventually the API has to convert the address to string to be used in iptables. IPv4 and IPv6 addresses are easy to differentiate, hence the different notation. Yes, with IPv6 some parts of the address can be omitted and should be treated as zeros, and there are IPv4-mapped IPv6 addresses to take care of. We can surely skip link local IPv6 and localhost addresses.

So it is a question about which part should do the conversion. The way I see that this plugin as an API is the place where input has to be verified and sanitized (also for leading white space) and not to rely on anything the application sends. Surely with string formatted addresses there is more work on the plugin. Other parameters can be changed to integers in next API version.

And a good point about the address with leading zeroes, these seem to pass through `getaddrinfo()` without errors. Need to check this.

LaakkonenJussi replied:
And also in https://tools.ietf.org/html/draft-main-ipaddr-text-rep-02 it is defined that: 

> 3.1 IPv4 Dotted Octet Format
>   A 32-bit IPv4 address is divided into four octets.  Each octet is represented numerically in decimal, using the minimum possible number of digits (leading zeroes are not used, except in the case of 0 itself).

So technically `012.0.0.1` is invalid input and should be treated as such.

LaakkonenJussi replied:
Added check of leading zeros in IPv4 addresses: f4aafb6587fb3396516ad38ffdbc4fa50d654dc5

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,464 @@
+ ip		s		IP address or network with CIDR or dot notation (1.2.3.4/24
+ 					or 1.2.3.4/255.255.255.0)
+ port		q		Port number as unsigned 16bit integer
+ port		s		Port as string, single port or port range, separator is ":"
```
I would suggest to encode port range as 2 numbers.

LaakkonenJussi replied:
Doable.

LaakkonenJussi replied:
Done in 113194ba7b5d66e0b1d2dfaa9fc53558c29b2ebf

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,464 @@
+ service	s		Service name, case insensitive, 
+ operation	s 		Perform operation add/remove, case insensitive, defaults to
+ 					add if empty string. Add appends the rule to iptables.
+ 					Remove removes the rule from iptables.
```
String? Case sensitive or not? Are leading/trailing spaces ignored or not? Just use a number, please. Numbers are so much easier to deal with.

monich replied:
This is not critical though. Feel free to ignore.

LaakkonenJussi replied:
See next comment. In convenience API methods I think I'll use integers to replace these strings.

LaakkonenJussi replied:
Changed operation to q (unsigned 16bit int) in 4df4a6dde2b8da11000867ed3a69aecc60ce3eee

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,464 @@
+<arg name="operation" type="s" direction="in"/>
+<arg name="result" type="q" direction="out"/>
+<arg name="string" type="s" direction="out"/>
+</method>
```
This could be (and probably is) the same as `AllowIncomingIpWithPort` with zero port. So it's unnecessary.

LaakkonenJussi replied:
Doable, maybe 4 different methods (allow in & out, deny in & out) could do the trick. I'll add that to next version with less string inputs.

LaakkonenJussi replied:
Methods are revamped in 113194ba7b5d66e0b1d2dfaa9fc53558c29b2ebf

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,464 @@
+<arg name="operation" type="s" direction="in"/>
+<arg name="result" type="q" direction="out"/>
+<arg name="string" type="s" direction="out"/>
+</method>
```
This is the same as `AllowIncomingIpWithPortRange` with the port range consisting of a single port, right? So it's unnecessary too.

LaakkonenJussi replied:
See prev. comment.

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,464 @@
+<arg name="operation" type="s" direction="in"/>
+<arg name="result" type="q" direction="out"/>
+<arg name="string" type="s" direction="out"/>
+</method>
```
Same thing as `AllowIncomingIp`.

LaakkonenJussi replied:
See prev. comment.

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,464 @@
+<arg name="operation" type="s" direction="in"/>
+<arg name="result" type="q" direction="out"/>
+<arg name="string" type="s" direction="out"/>
+</method>
```
Duplicate of `AllowIncomingIpWithPortRange`. Unnecessary.

LaakkonenJussi replied:
See prev. comment.

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,464 @@
+<signal name="FirewallSaved"></signal>
+
+<!-- Sent when ClearFirewall succeeds -->
+<signal name="FirewallCleared"></signal>
```
What's `ClearFirewall`?

LaakkonenJussi replied:
Changed method name in a636c6bfbd986ba333f506dbd593fdc470c13aa7 . this is sent when firewall is cleared. Maybe unnecessary.

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,464 @@
+<signal name="Shutdown"></signal>
+
+<!-- Sent when LoadFirewallFromDisk succeeds -->
+<signal name="FirewallLoaded"></signal>
```
`LoadFirewallFromDisk` is no more?

LaakkonenJussi replied:
Removed in a636c6bfbd986ba333f506dbd593fdc470c13aa7

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,464 @@
+<signal name="FirewallLoaded"></signal>
+
+<!-- Sent when SaveFirewallToDisk succeeds -->
+<signal name="FirewallSaved"></signal>
```
`SaveFirewallToDisk` is no more?

LaakkonenJussi replied:
Removed in a636c6bfbd986ba333f506dbd593fdc470c13aa7

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,464 @@
+"<!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN"
```
Leading quote?

LaakkonenJussi replied:
Removed in a636c6bfbd986ba333f506dbd593fdc470c13aa7

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,464 @@
+<arg name="operation" type="s"/>
+</signal>
+
+</interface></node>"
```
Stray quote?

LaakkonenJussi replied:
Fixed in a636c6bfbd986ba333f506dbd593fdc470c13aa7

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Added remove checks to test script.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Updated requirements in spec file.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Fix D-Bus message parameter use.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Improvements and restructuring to test script.

Added iptables as dependency for test script.

Also required commands are checked before running the test script.

Tests script now saves current iptables filter table if used as root. This is then restored when tests are succesfully executed.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Test script improvements; report errorous tests instead of exiting script and create log of all failed tests.

Changes to spec file descriptions.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Fixed connman dbus connection use.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Cleanup of API XML description, removing of signals that are not used and removing unused method and the related content. Improved validation of IPv4 and IPv6 addresses.

Added IPv6 address tests to unit tests.

LaakkonenJussi commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,464 @@
+<arg name="operation" type="s" direction="in"/>
+<arg name="result" type="q" direction="out"/>
+<arg name="string" type="s" direction="out"/>
+</method>
```
See prev. comment.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Corrected connman dbus connection use.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Implemented access control with libdbusaccess and signal unicasting.

Root user has access to all API functions. Group privileged can access all except clear iptables method. This will change in the future and this approach is only as a placeholder for future work (e.g., allow to manage rules per client).

Signals are now unicasted to all registered clients - clients have to use Register/Unregister D-Bus methods to gain/remove access to signals.

D-Bus methods can be used without registering if access is granted. Access denied return value is returned if the client does not have access to functions.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Removed unnecessary polkit .conf file and code cleanup. Added libdbusaccess and libglibutil as package dependencies.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Code restructuring: moved all policy functions to sailfish-iptables-policy.[ch]. Function renaming to achieve consistency.

Added unit tests for policy functions.

Fixes to issues found with unit tests and cleanup.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Changed unit test and test script permissions and updated descriptions and package names.

Improvements to test script. Users root, nemo, sailfish-mdm and privileged group permissions can be checked. Root can access all, nemo cannot access any parts of api and sailfish-mdm or privileged group can access all but clear. Checks for Register/Unregister but dbus-send quits after sending a message, thus, receiving unicast signals is not part of this test script.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Added protocol match (-m) to iptables rule spec used with connman to work with fixed iptables management implementation of connman.

Added requirement to recent connman version.

Added NULL entries as last in used charstring arrays.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Added QA testing xml and modified end-to-end test name and installation path.

For automated QA testing.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Updated documentation and small fixes to logging.

General cleanup of unneeded headers.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Added iptables save and restore test. Included in automated QA XML.

Script currently tests clearing of iptables and adding rules to iptables that they are saved correctly. Policy changes are not tested yet.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Restructured test scripts and moved general functionality to scripts under common.

Fixed most of the shellcheck errors and warnings in ete and save-restore test scripts.

Fixed file permissions of test-definition.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Added method for adding and removing (+ flushing) a chain to/from filter table.

Corrected unit tests and added test for chain adding and removal to ete test script.

Set "filter" as default table for all operations.

Connman require version bump.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Added a D-Bus method for returning iptables content of a specific table.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Fix QA test script.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Disable use of table as parameter over D-Bus for GetIptablesContent for now.

Enable this when other tables are required to be supported. Also for ClearIptablesTable.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Added a feature that records every chain add/removal internally and uses this list for removing custom chains from filter table at shutdown.

This feature cleans up the changes made to iptables filter table.

Increased required connman version.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Updated readme.md.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Added support for adding/removing custom chains for each iptables table, custom chains are prefixed (sfos_) and are removed by the plugin at shutdown - also restored at startup.

Updated readme and documentation.

Fixes for QA tests.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Added args and signal for clear chains method. Set to define chain name while processing parameters and to use it properly. Modified signals to have table names sent.

Modified also tests to check the removal of custom chains at service stop.

Added new common sources for tests (check and service) and modified tests accordingly.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Added loading of D-Bus policy from policy file.

Default policy is installed to /etc/connman/iptables_policy.conf. If file does not exist or is not accessible by the plugin default hardcoded D-Bus policy is used.

Added basic unit test as /etc/connman is not writable and custom config files cannot be tested.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Run end-to-end test script also as privileged user.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Added documentation package containing docbook xml format documentation (spec file modified to support docbook syntax).

Docbook XML is generated from spec xml. The "generate_html" script creates html page from the xml - this is not included in doc package as docbook5 is not available on sailfish.

Updated README.md accordingly.

Use makefile $(RM) instead of rm in Makefile cleanup.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Changed D-Bus method parameter operation from string to unsigned 16bit int.

Improved use of check_params() in all functions. Empty value for operation does no longer apply, approved values: 0 = add, 1 = remove, 2 = flush (only for ManageChain() method).

Documentation and tests updated accordgingly.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Cleanup of duplicate checks for values.

check_parameters() handles all cases to check for parameters.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Implemented 8 new D-Bus API methods that allow adding rules to any table and chain with a selected target name (iptables builtin targets or custom targets) and changing policy of a chain (INPUT, OUTPUT and FORWARD).

Each method has source and destination IPs and ports as input parameters. Source IP will be added with -s, destination IP with -d, source port with --sport and destination port with --dport. Also source and destination services are added, source service affects --sport and destination service affects --dport.

New methods have protocol set as unsigned 32bit integer that will be transformed to string using getprotobynumber(), unsigned 32bit int value corresponds to "all" protocol value.

Targets can be defined by hand, the builtin ACCEPT, DROP, REJECT, QUEUE, RETURN are supported and if the chain has been added via D-Bus API it can be used as a target. The custom chains have SAILFISH_IPTABLES_CHAIN_PREFIX which have to be omitted from the target name (API adds it).

Required connman version increased.

LaakkonenJussi committed:
Makefile to use makefile style rm command at cleanup.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Cleanup of old API.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Corrected order of D-Bus API parameters to correspond the order of the same parameters with iptables. Fixed save and restore test to use new API.

Readme and documentation spec file updated.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Add correct license to spec file.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Modify connman header name.

Upgraded connman requirement.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Removed saving/restore from plugin.

Saving and restoring is done by connman only.

Updated ete-test file.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Disable test_rule_post from tests.

Need to figure out the reason iptables is not cleared.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Add connman_iptables_content typedef

Change required by ConnMan changes in 1.32+git8.4.

Updated ConnMan version requirement.

monich commented on Makefile:
```diff
@@ -0,0 +1,131 @@
+CC = $(CROSS_COMPILE)gcc
+LD = $(CC)
+WARNINGS = -Wall
+BASE_FLAGS = -fPIC -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -Wformat -Wformat-security -fmessage-length=0
```
Please remove flags that are not absolutely needed. OBS will provide `CFLAGS` environment appropriate for the target platform. I think `-fPIC -fvisibility=hidden` should be enough.

LaakkonenJussi replied:
OK, will remove.

monich commented on README.md:
```diff
@@ -1,2 +1,278 @@
+
+When unloaded (by Connman at exit) iptables filter table is saved to connman's
+default save path, iptables content is cleared, custom chains are removed from
+the filter table and finally signal "Shutdown" is sent over D-Bus. Custom chains
```
This `Shutdown` signal seems to be useless (unlike `Initialize`). connman never unloads plugins. If the client needs to know when connman disappears from the bus, it would watch `net.connman` name - and be notified even if connman has crashed (in which case `Shutdown` signal won't be sent). I don't see any use case for this `Shutdown` thing.

LaakkonenJussi replied:
I find this necessary (both Shutdown and Initialize) as: the plugin may not be loaded because of version mismatch (highly unlikely but still possible), the plugin is registered/de-registered some time after connman is started or stopped that it is easier to follow the state than to send a message and get a reply telling that service does not exist and since this is the API 3rd parties are used who are not interested or should be using ConnMan directly they should have one place to listen to status changes.

monich replied:
I wasn't complaining about `Initialize`! `Initialize` is ok. My point was that listening for `Shutdown` is (currently) equivalent to watching the `net.connman` D-Bus name, except that watching the name is significantly more reliable. Whether or not 3rd parties are interested in using ConnMan directly, they still need this name, otherwise how would they find the service providing this interface.

monich replied:
But if you like it, leave it there! There's no harm in it.

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,490 @@
+</signal>
+
+
+</interface>
```
Never mind, I realized that (almost) each method has add/remove parameter. That seems slightly weird to me but I guess it's a matter of opinion.

monich replied:
I would have separate `AddRule` and `RemoveRule` methods but if you're going to maintain this stuff then it's up to you!

LaakkonenJussi replied:
I find less methods more suitable, more parameters. No need to duplicate that I think.

monich commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,490 @@
+	
+Sent when chain was added to or removed from a table.
+-->
+<signal name="ChainChanged">
```
Similar concern here. `ChainChanged` is not emitted when the chain is changed (i.e. rules are added or removed). But it is emitted when chains are being added or removed. I suggest separate `ChainAdded` and `ChainRemoved` signals.

LaakkonenJussi replied:
Chains can be added, removed and flushed. I'm not so keen on adding three different types of signals (one for each operation).

monich commented:
I was slightly confused by the new GitHub user interface, it could be that nobody was seeing the comments that I was typing :)

monich commented on src/sailfish-iptables-dbus.h:
```diff
@@ -0,0 +1,155 @@
+DBusMessage* sailfish_iptables_dbus_signal(const gchar* signal,
+	gint first_arg_type, ...);
+
+DBusMessage* sailfish_iptables_dbus_method_return(DBusMessage* message,
```
Couldn't find the implementation!

LaakkonenJussi replied:
Ah, maybe it had some use but now useless. Will remove.

LaakkonenJussi replied:
Changed in 961a9eb140971c7cb68e6dbea2febb140ab47427

monich commented on src/sailfish-iptables-dbus.h:
```diff
@@ -0,0 +1,155 @@
+
+/* These prototypes are connected to dbus */
+
+DBusMessage* sailfish_iptables_register_client(DBusConnection* connection,
```
This function is defined and used only by `sailfish-iptables-dbus.c`, it doesn't have to be external. It should be static.

LaakkonenJussi replied:
Then all of the functions connected to D-Bus are to be declared static.

LaakkonenJussi replied:
Will fix.

LaakkonenJussi replied:
Fixed in 961a9eb140971c7cb68e6dbea2febb140ab47427

monich commented on src/sailfish-iptables-dbus.h:
```diff
@@ -0,0 +1,155 @@
+DBusMessage* sailfish_iptables_register_client(DBusConnection* connection,
+			DBusMessage* message, void *user_data);
+			
+DBusMessage* sailfish_iptables_unregister_client(DBusConnection* connection,
```
This one too

LaakkonenJussi replied:
Will fix.

LaakkonenJussi replied:
Fixed in 961a9eb140971c7cb68e6dbea2febb140ab47427

monich commented on src/sailfish-iptables-dbus.h:
```diff
@@ -0,0 +1,155 @@
+DBusMessage* sailfish_iptables_unregister_client(DBusConnection* connection,
+			DBusMessage* message, void *user_data);
+
+DBusMessage* sailfish_iptables_clear_iptables_rules(DBusConnection *connection,
```
And this one

LaakkonenJussi replied:
Will fix.

LaakkonenJussi replied:
Fixed in 961a9eb140971c7cb68e6dbea2febb140ab47427

monich commented on src/sailfish-iptables-dbus.h:
```diff
@@ -0,0 +1,155 @@
+DBusMessage* sailfish_iptables_clear_iptables_rules(DBusConnection *connection,
+			DBusMessage *message, void *user_data);
+			
+DBusMessage* sailfish_iptables_clear_iptables_chains(DBusConnection *connection,
```
And this one

LaakkonenJussi replied:
Will fix.

LaakkonenJussi replied:
Fixed in 961a9eb140971c7cb68e6dbea2febb140ab47427

monich commented on src/sailfish-iptables-dbus.h:
```diff
@@ -0,0 +1,155 @@
+DBusMessage* sailfish_iptables_clear_iptables_chains(DBusConnection *connection,
+			DBusMessage *message, void *user_data);
+			
+DBusMessage* sailfish_iptables_get_iptables_content(DBusConnection *connection,
```
And this one

LaakkonenJussi replied:
Will fix.

LaakkonenJussi replied:
Fixed in 961a9eb140971c7cb68e6dbea2febb140ab47427

monich commented on src/sailfish-iptables-dbus.h:
```diff
@@ -0,0 +1,155 @@
+DBusMessage* sailfish_iptables_get_iptables_content(DBusConnection *connection,
+			DBusMessage *message, void *user_data);
+
+DBusMessage* sailfish_iptables_version(DBusConnection *connection,
```
And this one

LaakkonenJussi replied:
Will fix.

LaakkonenJussi replied:
Fixed in 961a9eb140971c7cb68e6dbea2febb140ab47427

monich commented on src/sailfish-iptables-dbus.h:
```diff
@@ -0,0 +1,155 @@
+DBusMessage* sailfish_iptables_version(DBusConnection *connection,
+			DBusMessage *message, void *user_data);
+
+DBusMessage* sailfish_iptables_change_policy(
```
And this one

LaakkonenJussi replied:
Will fix.

LaakkonenJussi replied:
Fixed in 961a9eb140971c7cb68e6dbea2febb140ab47427

monich commented on src/sailfish-iptables-dbus.h:
```diff
@@ -0,0 +1,155 @@
+#define SAILFISH_IPTABLES_SIGNAL_CHAIN		"ChainChanged"
+
+#ifdef __cplusplus
+extern "C" {
```
How this header could possibly be included from C++?

LaakkonenJussi replied:
Copy-paste-error, will fix.

LaakkonenJussi replied:
Fixed in 961a9eb140971c7cb68e6dbea2febb140ab47427

LaakkonenJussi commented on rpm/sailfish-iptables-api-plugin.spec:
```diff
@@ -0,0 +1,123 @@
+%files
+%defattr(-,root,root,-)
+%{_libdir}/connman/plugins/sailfish-connman-iptables-plugin.so
+%config %{_sysconfdir}/connman/*.conf
```
OK, will change.

LaakkonenJussi replied:
Changed in 961a9eb140971c7cb68e6dbea2febb140ab47427

LaakkonenJussi commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,490 @@
+	
+	Sends signal RuleChanged.
+-->
+<method name="RuleIpWithService">
```
Well this is a easy to use method, no big harm having it either I think.

LaakkonenJussi commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,490 @@
+	@result: result code, 0 = Request done, 8 = table name was invalid, 9 = cannot make change (iptables error), 100 = access denied
+	@string: result string.
+	@chains: list of chains in the table, format: CHAIN POLICY
+	@rules: list of rules in the table, format is similar to iptables-save command format
```
No, it is not binary form. The rules in iptables are in text format.

LaakkonenJussi replied:
Similar to iptables-save format means that the rule part is unchanged when comparing to iptables-save format.

LaakkonenJussi commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,490 @@
+	
+Sent after a rule is succesfully added to or removed from iptables. Any of the string parameters can be empty "". If port is 0 then port is not added with the rule. If destination/source end port is 0 then port range is not utilized and the ports defined in destination/source start is only used.
+-->
+<signal name="RuleChanged">
```
In iptables rules are only added or removed. Change would mean that remove Nth rule from a chain and add a rule to same position as changed rule. 2 operations, remove and add (actually insert). I'll change the signals to be more descriptive (add and remove).

LaakkonenJussi replied:
Changed in 961a9eb140971c7cb68e6dbea2febb140ab47427

LaakkonenJussi commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,490 @@
+	
+	Sent when plugin is removed from Connection Manager.
+-->
+<signal name="Shutdown"></signal>
```
I'm not seeing it as unnecessary - as I said, I prefer to track the API not ConnMan. Although the API is for ConnMan I really do not see the point to complicate things with multiple points of interest.

LaakkonenJussi commented on spec/sailfish_iptables_dbus_interface_description.xml:
```diff
@@ -0,0 +1,490 @@
+	ClearIptablesTable:
+	@table: Iptables table to clear.
+	@result: result code, 0 = cleared, 8 = table name was invalid, 9 = cannot make change (iptables error), 100 = access denied
+	@string: result string.
```
Well this was more of debug purposes. Maybe it'll result less work in the future. I'll change this to return the code only.

LaakkonenJussi replied:
Changed in 961a9eb140971c7cb68e6dbea2febb140ab47427

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Cleanup, test fix and new signals.

This commit contains:
 - Reduced amount of compile flags as OBS sets them automatically.
 - Split RuleChanged signal to RuleAdded and RuleRemoved. These signals
   omit the operation type as parameter. Spec and docs updated.
 - ConnMan version requirement bump to 1.32+git9.
 - Added configuration package for plugin configs.
 - Removed descriptive string from return message, only code will be
   returned as reply. Spec, docs and tests updated.
 - All functions connected to D-Bus are declared now static.
 - Fixed save-restore-test post cleanup test - test failed because there
   was a reference to custom chain in some other chain.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Cleanup of license text and code.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Cleanup rule adding function.

Group cases together to make code more manageable.

monich commented on src/sailfish-iptables-parameters.h:
```diff
@@ -0,0 +1,110 @@
+void rule_params_set_protocol_str(rule_params *params, const gchar *protocol);
+void rule_params_set_protocol(rule_params *params, guint16 protocol);
+void rule_params_set_ip_src_negate(rule_params *params, gboolean value);
+void rule_params_set_ip_dst_negate(rule_params *params, gboolean value);
```
The above `rule_params_set_*` functions don't seem to be used anywhere

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Version bump to 0.0.4 and cleanup.

Removed unused function prototypes and moved D-Bus method names to
header.

monich commented on src/sailfish-iptables-validate.c:
```diff
@@ -0,0 +1,386 @@
+	gint length = 0, i = 0;
+	
+	if(!address && !*address)
+		return false;
```
A strange mix of `gboolean` from glib and `true`/`false` from `<stdbool.h>`. glib  comes with `gboolean`/`TRUE`/`FALSE` and `<stdbool.h>` with `bool`/`true`/`false`. Either one is ok but why mix those? Not a big deal of course.

LaakkonenJussi replied:
Then I remembered wrong, I thought it was the exact opposite.

LaakkonenJussi replied:
I think these kind of things can be fixed later on.

LaakkonenJussi committed:
[sailfish-connman-iptables-plugin] Documentation and spec update.

LaakkonenJussi commented on src/sailfish-iptables-dbus.h:
```diff
@@ -0,0 +1,97 @@
+#define SAILFISH_IPTABLES_SIGNAL_POLICY		"PolicyChanged"
+#define SAILFISH_IPTABLES_SIGNAL_RULE_ADD		"RuleAdded"
+#define SAILFISH_IPTABLES_SIGNAL_RULE_REM		"RuleRemoved"
+#define SAILFISH_IPTABLES_SIGNAL_CHAIN		"ChainChanged"
```
Moved in ee72244f628c27b42d0539eb49f8b633756cb9a6
